### DISCUSSION TITLE
how to create text backgrounds like tiktok

### DISCUSSION THREAD

**Discussion by andykais** [ImageMagick/ImageMagick#3090] [Help]
Hi, is it possible to create text backgrounds like those found on tiktoks? Here is an example:

![gordon-ramsay-tiktok-reactions-png-1596729906](https://user-images.githubusercontent.com/6627147/103918366-b7b43580-50dc-11eb-961a-57b4c8eb35a9.png)

with this command:
```bash
magick -size 350x -font DejaVu-Sans -background white -gravity Center -pointsize 30 -fill black -stroke none 'caption:Genuinely so nice and simple' out.png
```
I have gotten this far 
![out](https://user-images.githubusercontent.com/6627147/103919992-bb48bc00-50de-11eb-80b8-487dbea9b1df.png)

but it does not contour the edges of the text, and does not round the edges. Any help is appreciated!

**Comment by fmw42** [comment]
You just need to add the background image and swap to make the background image first, then do -compose over -composite

```
magick -size 350x -font DejaVu-Sans -background white -gravity Center -pointsize 30 -fill black \
-stroke none 'caption:Genuinely so nice and simple' \
barn.jpg +swap -compose over -composite out.png

```

![out](https://user-images.githubusercontent.com/13458218/103928838-3e631580-50d1-11eb-85ba-73f3a8567e6a.png)

This is Unix syntax. For windows replace the \ with ^

**Reply by andykais** [reply]
I dont think this answers my question. I am not concerned that I cannot add a background image. I want to know how I can make the white background contour the text similar to the first image . Note I do not want to outline the text, I want a rectangular border that wraps along with the words

**Comment by andykais** [comment]
got a little further with `-undercolor`! I still dont have rounded edges though. Im not sure how I can add that :thinking: 
```bash
magick -size 350x -font DejaVu-Sans -background none  -undercolor white -gravity Center -pointsize 30 -fill black -stroke none 'caption:Genuinely so nice and simple' out.png
```
![out](https://user-images.githubusercontent.com/6627147/103939981-a5e18b00-50fa-11eb-89cd-2ccf96e190b8.png)

**Comment by fmw42** [comment]
```
magick -size 400x -font DejaVu-Sans -background black -undercolor white \
-gravity Center -pointsize 30 -fill white -stroke none "caption:Genuinely so nice and simple" \
-shave 2x2 -bordercolor black -border 10 -border 2x2 \
-background white -virtual-pixel white -blur 0x3 -threshold 50% -alpha off -shave 2x2 \
\( -size 350x -fill red -stroke none "caption:Genuinely so nice and simple"  \) \
-alpha off -auto-level -compose multiply -composite -compose over \
-fuzz 20% -fill none -draw "alpha 0,0 floodfill" \
-fuzz 80% -fill black -opaque red \
barn.jpg +swap -compose over -composite result.png

```
If you are on Window, then try the following

```
 magick -size 400x -font DejaVu-Sans -background black -undercolor white ^
-gravity Center -pointsize 30 -fill white -stroke none "caption:Genuinely so nice and simple" ^
-shave 2x2 -bordercolor black -border 10 -border 2x2 ^
-background white -virtual-pixel white -blur 0x3 -threshold 50% -alpha off -shave 2x2 ^
( -size 350x -fill red -stroke none "caption:Genuinely so nice and simple" ) ^
-alpha off -auto-level -compose multiply -composite -compose over ^
-fuzz 20% -fill none -draw "alpha 0,0 floodfill" ^
-fuzz 80% -fill black -opaque red ^
barn.jpg +swap -compose over -composite result.png

```

![result](https://user-images.githubusercontent.com/13458218/103976926-e6f29300-512c-11eb-8d51-f909a36d0b07.png)

**Comment by GeeMack** [comment]
You can get the effect of expanding the shape of the text area using ImageMagick's "-morphology" operation. This command is the same as you had up to making the caption label, then continues at "-bordercolor none"...

    magick -size 350x -font DejaVu-Sans -background none \
       -undercolor white -gravity Center -pointsize 30 -fill black \
       -stroke none 'caption:Genuinely so nice and simple' \
       -bordercolor none -border 12 \
       \( +clone -morphology dilate disk:8x8 -blur 0x2 \) \
       +swap -composite barn.jpg +swap -composite out.png

After making the label the command adds a transparent border. Then inside the parentheses it clones the label and uses a morphology kernel "disk" to expand the clone and make it with rounded corners. I put a blur inside the parentheses, too.

Then after the parentheses it composites the text label onto that rounded-corner, expanded, blurred background. To finish, read in the main image, "+swap" it with the caption label, and composite the label onto the main image.

![out](https://user-images.githubusercontent.com/65475279/103979449-6f7c2e00-5143-11eb-8168-13939799931a.png)

**Reply by andykais** [reply]
this is getting pretty close! Building off of your command, I have this:
```bash
magick -size 350x -font DejaVu-Sans -background none \
   -undercolor white -gravity Center -pointsize 30 -fill black \
   -stroke none 'caption:Genuinely so nice and simple' \
   -bordercolor none -border 12 \
   \( +clone -morphology dilate disk:4 \) \
   +swap -composite out.png           
```

![out](https://user-images.githubusercontent.com/6627147/104035213-de877000-519f-11eb-8013-5b5cdd4e0372.png)

Im going to be a stickler though :sweat_smile: and say that the width of the background is a little shorter than I would like. Compare that to the expected output here:
![gordon-ramsay-tiktok-reactions-png-1596729906](https://user-images.githubusercontent.com/6627147/103918366-b7b43580-50dc-11eb-961a-57b4c8eb35a9.png)

 I dont quite know how to fix this since it seems like the [disk](https://legacy.imagemagick.org/Usage/morphology/#disk) kernel only accepts a single param radius, so in your example it only takes the first `8` out of `8x8`. I think what I want is some sort of elipse kernel, but I dont see one of those available as a built in

[edit] the inset corners are also hard edges. Is there any way to also round those corners?

**Reply by andykais** [reply]
I managed to get what I want with an _obnoxiously_ large custom dilation command:
![out](https://user-images.githubusercontent.com/6627147/104046493-2e6d3380-51ae-11eb-9994-06db20d76ec6.png)

```bash
magick -size 350x -font DejaVu-Sans -background none -undercolor white -gravity Center -pointsize 30 -fill black -stroke none 'caption:Genuinely so nice and simple' -bordercolor none -border 12 '(' +clone -morphology dilate '21x7: 
nan,nan,nan,1,nan,nan,nan
nan,nan,nan,1,nan,nan,nan
nan,nan,nan,1,nan,nan,nan
nan,nan,1,1,1,nan,nan
nan,nan,1,1,1,nan,nan
nan,nan,1,1,1,nan,nan
nan,1,1,1,1,1,nan
nan,1,1,1,1,1,nan
nan,1,1,1,1,1,nan
1,1,1,1,1,1,1
1,1,1,1,1,1,1
1,1,1,1,1,1,1
nan,1,1,1,1,1,nan
nan,1,1,1,1,1,nan
nan,1,1,1,1,1,nan
nan,nan,1,1,1,nan,nan
nan,nan,1,1,1,nan,nan
nan,nan,1,1,1,nan,nan
nan,nan,nan,1,nan,nan,nan
nan,nan,nan,1,nan,nan,nan
nan,nan,nan,1,nan,nan,nan' ')' +swap -composite out.png
```

The dither shape looks like this:
```
...0...
...0...
...0...
..000..
..000..
..000..
.00000.
.00000.
.00000.
0000000
0000000
0000000
.00000.
.00000.
.00000.
..000..
..000..
..000..
...0...
...0...
```

Im really hoping theres a builtin solution though. The rounding is still not great, and this is going to be used in a library. I want to allow the rounding to be configurable. I really dont want to have to write a custom dither shape calculator based on a rounding value. This was 
more of a "see if it was possible" rather than an actual solution. Im still not able to round the inset corners either

**Comment by fmw42** [comment]
See https://legacy.imagemagick.org/Usage/morphology/#disk It appears that the disk kernel does not permit two values. However you can use the DIY kernel that has the same properties as the disk and set it up for two values.

With the following code one can define the white size (-border X) and the rounding (-morphology dilate disk:5)


small rounding
```
magick -size 350x -font DejaVu-Sans -background none \
-undercolor white -gravity Center -pointsize 30 -fill black \
-stroke none caption:"Genuinely so nice and simple" \
-bordercolor none -border 12 \
\( +clone -morphology dilate disk:5 \) \
+swap -compose over -composite barn.jpg +swap -composite result6.png

```

![result6](https://user-images.githubusercontent.com/13458218/104078098-430ef300-51d1-11eb-8cf5-6332e53cfac2.png)


larger rounding
```
magick -size 350x -font DejaVu-Sans -background none \
-undercolor white -gravity Center -pointsize 30 -fill black \
-stroke none caption:"Genuinely so nice and simple" \
-bordercolor none -border 12 \
\( +clone -morphology dilate disk:15 \) \
+swap -compose over -composite barn.jpg +swap -composite result7.png
```


![result7](https://user-images.githubusercontent.com/13458218/104078103-4a360100-51d1-11eb-928b-94e4ffc113bd.png)

**Reply by nusantero** [reply]
Tried this on Raspberry Pi, not working as expected.

**Reply by snibgo** [reply]
If you want any help, you need to tell us more about the problem.

What version of IM? Is "Raspbery Pi" syntax like bash?

What command did you try? What was your input image?

What went wrong? What did you expect, and what did you get instead?

**Comment by snibgo** [comment]
For clarity, I use a blue background #44f.

The rectangular outline can be made from "-undercolor". The border must be sufficiently large for dilation etc. We extract the alpha for further processing. We dilate by a rectangle. The dimensions of the rectangle give the expansion horizontally and vertically.

Then you want to round all corners, with a different radius horizontally and vertically. We do this by blurring horizontally then vertically, and threshold the result. This gives rounded corners. A small blur blurs the edges.

There are essentially five tweakable parameters: expansion in x and y, rounding in x and y, and final blur. I save intermediate results x0.png etc so you can see the individual effects.

Windows BAT script:
```
magick ^
  -size 350x -undercolor #44f -gravity Center -pointsize 30 ^
  -background None ^
  caption:"Genuinely so nice and simple" ^
  -bordercolor None -border 20 ^
  -background #44f -alpha background ^
+write x0.png ^
  ( +clone ^
    -alpha extract ^
+write x1.png ^
    -morphology Dilate Rectangle:13x7 ^
+write x1a.png ^
    -morphology Convolve Blur:0x6 ^
    -morphology Convolve Blur:0x3,90 ^
    -threshold 50%% ^
    -blur 0x2 ^
+write x2.png ^
  ) ^
  -alpha off -compose CopyOpacity -composite ^
  -trim +repage ^
  x.png
```
![x](https://user-images.githubusercontent.com/33812368/104093771-a017b100-5284-11eb-88cd-15c857eb8794.png)

**Reply by never-failing** [reply]
Please tell me how to make the background transparent. I tried to add -transparent-color instead of -undercolor, the background was completely gone ![Screenshot_17](https://user-images.githubusercontent.com/76694308/199030613-28e0236d-844a-4cbe-bd62-038e037a9558.png)
Here's my command `magick -size 900x -undercolor #FFFFFF -gravity Center -fill #62639B -font AdigianaUltra.ttf -pointsize 74 -background None caption:"КОМПОТ ИЗ ЧЁРНОЙ СМОРОДИНЫ С АПЕЛЬСИНОМ НА ЗИМУ" -bordercolor None -border 40 -background #FFFFFF -alpha background +write x.png ( +clone -alpha extract +write x.png -morphology Dilate Rectangle:13x7 +write x.png -morphology Convolve Blur:0x6 -morphology Convolve Blur:0x3,90 -threshold 50%% -blur 0x2 +write x.png ) -alpha off -compose CopyOpacity -composite -trim +repage x.png`

**Comment by fmw42** [comment]
Generally, to make some color transparent, use `-fuzz XX% -transparent some-color`. In this case you want to make everything but some color transparent. So use `-fuzz XX% +transparent some-color`, where some-color is your text color.

`magick 199030613-28e0236d-844a-4cbe-bd62-038e037a9558.png -alpha off -fuzz 5% +transparent "rgb(97,99,158)" result.png`

![result](https://user-images.githubusercontent.com/13458218/199045973-94280146-9422-4b89-a644-a4610ff47d4a.png)

**Reply by never-failing** [reply]
I'm sorry, I probably didn't explain it well. I need a transparent background like in the screenshot below
![Screenshot_18](https://user-images.githubusercontent.com/76694308/199180993-772567d4-eaa0-429a-9ae3-aeca5be0a358.png)

**Comment by fmw42** [comment]
Here is one way that is close.  Basically, we create a second version of the image that has been darkened. Then we create a mask by extracting the alpha channel from the non-blur letters and then using morphology to spread out the white part to fill out the letters in to a complete single region rather than a region for each letter. Then subtract the letters from the larger region.  This mask is then used to blend the original and darkened image.

![199030613-28e0236d-844a-4cbe-bd62-038e037a9558](https://user-images.githubusercontent.com/13458218/199333033-c3d5264e-d3fd-48af-9466-69024d1dcb0d.png)

Unix syntax:

```
magick 199030613-28e0236d-844a-4cbe-bd62-038e037a9558.png -write mpr:img \
\
\( mpr:img -fill black -colorize 25% \) \
\
\( mpr:img -fuzz 5% +transparent "rgb(97,99,158)" \
-alpha extract +write mpr:alpha \
-bordercolor black -border 50 \
-morphology dilate rectangle:50x50 \
-morphology erode rectangle:40x30 \
-shave 50x50 \
mpr:alpha -compose difference -composite \) \
\
-compose over -composite \
result.png
```

![result](https://user-images.githubusercontent.com/13458218/199333198-4c02f382-1c7e-41f0-8467-265afd5e9fe2.png)

**Comment by phoenix-gi** [comment]
Here a simple solution to place on image text with rounded border:
```sh
magick convert input.jpg \( \( -background blue -fill white -font Arial -pointsize 32 label:'hello, world!' \) -bordercolor blue -border 16 -alpha set -virtual-pixel transparent -channel A -blur 16x16  -threshold 50% +channel \) -gravity center -composite output.jpg
```
- You can manipulate the background color by changing `-background blue` and `-bordercolor blue` parameters
- You can manipulate text color by changing `-fill white` parameter
- You can manipulate the border radius by changing `-border 16` and `-blur 16x16` parameters
- You can manipulate text size by changing `-pointsize 32` parameter
- To move the text somewhere else, you can change last `-gravity center` parameter

How it works:
- First it generates text image with background by command:
  - `-background blue -fill white -font Arial -pointsize 32 label:'hello, world!'`
- Then it applies to the generated text image following command to make rounded border:
  - `-bordercolor blue -border 16 -alpha set -virtual-pixel transparent -channel A -blur 16x16  -threshold 50% +channel`
- Finally it applies `-gravity center -composite` to place rounded bordered text on the center of `input.jpg` image

I hope it will be useful for anyone! Thanks.
